### ISSUE TITLE
strange error output on recursive function calls

### ISSUE THREAD

**Issue by LukasBanana** [Microsoft/DirectXShaderCompiler#62]
Recursive calls are not allowed in shading languages, but the compiler should report a meaningful error in that case.

I use the following example code with the online HLSL compiler ([tryhlsl.azurewebsites.net](http://tryhlsl.azurewebsites.net/)):
```c
float4 f(); // Forward declaration
float4 g() { return f(); } // Recursive call to 'f'
float4 f() { return g(); } // First call to 'g'

float4 VS() : SV_Position
{
    return f(); // First call to 'f'
}
```
The error output of the old HLSL compiler with target "vs_5_1" is:
```
D:\Windows\system32\unknown(4,8-10): error X3500: 'f': recursive functions not allowed in vs_5_1
```

But the error output of the new HLSL compiler with target "vs_6_0" is:
```
error: validation errors
Function ?f@@YA?AV?$vector@M$03@@XZ.flat with parameter is not permitted, it should be inlined
Not all elements of output SV_Position were written
Not all elements of SV_Position were written
```

I assume that this is not the intended output.

**Comment by dmpots** [comment]
Yes, the error message should definitely be better here. Thanks for the report!

**Comment by marcelolr** [comment]
Just an update, I'm working on this but it's a bit tricky. If we do this in the front-end right away, where we can provide the best diagnostics, we run into an inability to support this.

```
static int i = 10;

float4 f(); // Forward declaration
float4 g() { if (i > 10) { i--; return f(); } else return 0; } // Recursive call to 'f'
float4 f() { return g(); } // First call to 'g'

float4 VS() : SV_Position {
    return f(); // First call to 'f'
}

```
This works in fxc, which will inline while tracking the value of 'i' and actually collapse the shader into a simple constant output. We can best approximate this by doing the validation for recursion once we're processing DXIL, but it might be harder to recove all proper information for diagnostics at that point (but don't lose hope, I'm just getting started on this).

Long-term, this is one of those things we'd like to fix by revising the language so we have better defined rules as to what programs are valid from a language perspective and what behavior you can expect from your compiler (and presumably the proper mechanisms to control the latter to get what you want).

**Comment by LukasBanana** [comment]
So you mean dxc should support your code example without a compiler error?
But doesn't the fact, that your example is valid for _fxc_, rely on the compiler optimization passes?!
Without optimization there is still an invalid recursive function call, even if it's _dead code_, due to the always-false-condition.
However, I think the compiler should always reject such code, no matter if a condition will ever happen,
since recursion is not allowed at all in shading languages.

**Comment by marcelolr** [comment]
Fair enough. We're trying to keep this first release as back-compat friendly as possible to make it easier for developers to adopt it, and we have seen many instances were developers take advantage of fxc's optimizations to write code that cannot be translated to valid bytecode without the compiler cleaning things up.

This does seem pretty out there, so hopefully this kind of recursion isn't something we'll run into in the wild; I'll implement a check earlier on and keep an eye out for cases.

Hopefully in the next iteration of the language we'll have a mechanism to do something along these lines that doesn't rely on optimization levels (if we introduce function templates, I'm sure someone will use recursion on template instantiation for this...)
